src/visualizers/Chaos.ts

```diff
@@ -376,8 +384,8 @@ class Chaos extends VisualizerDefault {
         this.sketch.strokeWeight(0)
```

Comment:
Can you explain the modifications to the visualizers a bit more?  I've noticed many changes are referring to `sketch` not `this.sketch`, but not everywhere (e.g. here).  It looks like here there's some issue that `sketch` isn't working quite yet so we need to fall back?

Reply:
OK, this is a bit of a long story, sorry. (Too bad we don't have time left for a meeting on this.) We were never exercising the p5 event handling, as you know, because it didn't work because of the way Numberscope had structured the interaction between the Vue framework and the visualizer code. So when I turned that on, events were indeed being sensed, but they were all being executed multiple times, and the more I played with the system, the more and more times each event was being processed. Finally I figured out that every time you hit a "Draw" button on a BundleCard and then every time you hit the "<Back" button on the main drawing page, a new p5 sketch was being created, which isn't so bad, but all of the old p5 sketches, even though they were no longer in the active DOM, were still running and plotting themselves and doing everything including processing events, but just not being displayed because they were no longer in the DOM.

So with the current design of the  Vue framework, I had no choice but to add methods to dispose of the old sketch instances when what's active in the DOM changes. But that in turn means that the type of the `sketch` member of a P5Visualizer **had** to change from `p5` to `p5 | undefined`, because it truly may be the case that for some stretch of a visualizer's lifetime, after it has been created and maybe after it at one time had a valid sketch, it might not have any sketch (if the framework has temporarily shelved it with `dispose()` but might later revive it by having it `inhabit()` a new DOM element, when it rearranges the screen.

Therefore, typescript wants all the code to deal with the possibility that `this.sketch` might be undefined. And it is basically right, except for the fact that for most of the member functions, at the point they are called, there really will be a valid sketch object in the `this.sketch` property. But to keep the TypeScript compiler happy (and, frankly, to be really sure the code won't under any circumstances barf), all the functions need to be sure to work when this.sketch comes in undefined.

The two sort of special cases here are `setup()` and `draw()`:
  (A) When `setup` is entered, this.sketch() **is** always undefined -- it is setup's responsibility to _create_ the sketch. But that happens in P5Visualizer.setup(), and all derived p5 visualizers are required to call super.setup() so that the sketch will be initialized. So after that call to super.setup(), this.sketch is guaranteed to be valid. I wasted an hour or so trying to get TypeScript to realize that once super.setup() returns successfully, this.sketch is automatically valid, but I failed to find a way to do so. That's why you see all derived setup() functions (redundantly) discarding the possibility that this.sketch might be undefined, just after calling super.setup(). If anyone has any ideas how to smooth this, I am all ears -- I did try.

  (B) When `draw` is entered, this.sketch is absolutely guaranteed to be valid -- with the design of p5 and the P5Visualizer class, there is no way `draw()` can be called without a valid sketch, because it is the sketch itself that is calling draw! But TypeScript has no way of knowing this. Moreover, draw is the place where 90% of the uses of this.sketch occur anyway. So I compromised and changed the structure of P5Visualizer so that it would pass in that certainly valid sketch as the `sketch` argument to draw. It is guaranteed to be equal to this.sketch, and it is guaranteed to be a valid p5 sketch so you don't have to runtime typecheck the sketch argument to draw to make TypeScript happy, like you would the this.sketch property. And it also seemed like a bit of a win because since the big bulk of sketch references occur in draw, so being able to refer to the sketch just as the `sketch` argument rather than the `this.sketch` property also allowed removing dozens of occurrences of `this.` from the code, making it more concise. So the upshot -- which I grant may be a bit confusing -- is that the `sketch` argument to `draw()` **is guaranteed** to be `this.sketch`, just pre-type-certified as valid for the TypeScript compiler.

Again, if you want to revisit this design/handling of the sketch object in the P5Visualizer hierarchy, I am super happy to if anyone can come up with a smoother system. I just couldn't see any alternative to allowing the possibility in the code that this.sketch might be undefined, because the framework really does need the ability to throw away a sketch after it has done some amount of drawing. And so the visualizer has no way of knowing for sure at any point in its lifetime that this.sketch is currently valid, as far as I could see.

Reply:
Holy Cow!  Ok.  I'm actually glad this is in writing, not just in a zoom meeting.  I suggest @Vectornaut  link to this comment or discussion as a code comment in the example visualizer code he will be adding, so that when someone else wonders, they can dive into it if they care, and so they don't try to circumvent it or get confused.  If there's anywhere appropriate to add a link or explanation in a comment inside the code in this PR, that would be great.  Otherwise, I think it is what it is.

Reply:
Ah actually in thinking about this situation more in the context of your Grid.ts comment a few conversations below, I did come up with a couple of other design tweaks that seems like it could streamline this considerably. So before responding here, do take a look at the options there and let me know.

Reply:
Yeah, that was another example of conversation overlap. Will be interested to hear what you think about the new proposal.